File: 17-trees3/3-tree-serialization/main.py
# Tree Serialization - Python Example

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def serialize(root):
    def helper(node):
        if node is None:
            return "#,"
        return str(node.val) + "," + helper(node.left) + helper(node.right)
    return helper(root)

def deserialize(data):
    values = iter(data.split(','))
    
    def helper():
        val = next(values)
        if val == "#":
            return None
        node = TreeNode(int(val))
        node.left = helper()
        node.right = helper()
        return node
    return helper()

# Usage example
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.right.left = TreeNode(4)
root.right.right = TreeNode(5)

serialized = serialize(root)
print("Serialized tree:", serialized)
new_root = deserialize(serialized)
print("Deserialized root value:", new_root.val)

File: 17-trees3/3-tree-serialization/main.java
// Tree Serialization and Deserialization - Java Example

import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int val) {
        this.val = val;
    }
}

class TreeSerialization {

    // Serializes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) return "#,";
        return root.val + "," + serialize(root.left) + serialize(root.right);
    }

    // Deserializes encoded data to tree.
    public TreeNode deserialize(Queue<String> nodes) {
        String val = nodes.poll();
        if (val.equals("#")) return null;
        TreeNode node = new TreeNode(Integer.parseInt(val));
        node.left = deserialize(nodes);
        node.right = deserialize(nodes);
        return node;
    }

    public TreeNode deserialize(String data) {
        String[] parts = data.split(",");
        Queue<String> nodes = new LinkedList<>(Arrays.asList(parts));
        return deserialize(nodes);
    }

    public static void main(String[] args) {
        TreeSerialization codec = new TreeSerialization();

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(5);

        String serialized = codec.serialize(root);
        System.out.println("Serialized tree: " + serialized);

        TreeNode newRoot = codec.deserialize(serialized);
        System.out.println("Deserialized root value: " + newRoot.val);
    }
}
